### PR TITLE
Guidelines for delta audits and `config.toml` policy.

### PR DESCRIPTION
**Pull request by anforowicz**

### CODE REVIEW

File: auditing_standards.md [hunk 1]
```diff
@@ -237,3 +237,73 @@ Crates with this criteria contain very dangerous unsafe rust code.
     *   Undefined behavior may be possible under common circumstances.
 *   Most crates that try to be sound but don't quite make the cut go in `ub-risk-3`, not here. These crates are wildly unsafe, and the only time we should use them is when they are a necessary evil.
 *   Everything worse than `ub-risk-3` goes in here and we should do our best to avoid using them.
+
+## Criteria-agnostic guidelines
+
+### Delta audits should describe the final version
+
+In general, delta audits must never lower the `UB-risk-N` value from the
```
Comment (Manishearth) [inline auditing_standards.md:245]: I disagree: delta audits can lower the UB risk if the delta fixed _all_ previous delta issues.

I think we should say that a review performed of _just_ a delta should never lower the risk. The risk may be lowered by looking at the delta _and_ the issues found in the previous review.

Comment (anforowicz) [reply auditing_standards.md]: Good catch - makes sense.  I tried to reword this paragraph to capture your point.

File: auditing_standards.md [hunk 2]
```diff
@@ -237,3 +237,76 @@ Crates with this criteria contain very dangerous unsafe rust code.
     *   Undefined behavior may be possible under common circumstances.
 *   Most crates that try to be sound but don't quite make the cut go in `ub-risk-3`, not here. These crates are wildly unsafe, and the only time we should use them is when they are a necessary evil.
 *   Everything worse than `ub-risk-3` goes in here and we should do our best to avoid using them.
+
+## Criteria-agnostic guidelines
+
+### Delta audits should describe the final version
```
Comment (djkoloski) [inline auditing_standards.md:243]: This change assumes that doing this:

1. Audit crate version 1 with ub-risk-3.
2. Audit crate delta 1->2 with ub-risk-0.

Results in crate version 2 having ub-risk-0. To my mind, this series of events should still result in version 2 having ub-risk-3, but I may understand how the audits are interacting. As I understand it, delta audits can only ever remove audited criteria by [acting as non-traversable edges](https://github.com/mozilla/cargo-vet/blob/main/src/resolver.rs#L1598-L1601). If that is the case, then we should **not** require delta audits to describe the final state of the crate.

File: auditing_standards.md [hunk 3]
```diff
@@ -237,3 +237,76 @@ Crates with this criteria contain very dangerous unsafe rust code.
...
+[`lib.rs` on line 1282](https://chromium-review.googlesource.com/c/chromium/src/+/5453972/2..6/third_party/rust/chromium_crates_io/vendor/proc-macro2-1.0.80/src/lib.rs#1282)).
+
+### Audits should ignore criteria required by `config.toml` policy
+
+Audits should record the actual risk of the crate, even if a looser
+certification would be sufficient to pass a `config.toml` policy.
+
+For example, Chromium requires `ub-risk-2` for crates used in the browser
+process - e.g. see the following auto-generated `config.toml`
+[entry for the `quote` crate](https://source.chromium.org/chromium/chromium/src/+/main:third_party/rust/chromium_crates_io/supply-chain/config.toml;l=158-159;drc=aaf3f1dab40ed4248391227583a1590bfbf7062e):
```
Comment (djkoloski) [inline auditing_standards.md:265]: Nit: remove link, all of the relevant text is included inline so we should reduce our exposure to link rot.
```suggestion
entry for the `quote` crate:
```

File: auditing_standards.md [hunk 4]
```diff
@@ -237,3 +237,76 @@ Crates with this criteria contain very dangerous unsafe rust code.
...
+(e.g. `unsafe { imp::Literal::from_str_unchecked(repr) }` in
+[`lib.rs` on line 1282](https://chromium-review.googlesource.com/c/chromium/src/+/5453972/2..6/third_party/rust/chromium_crates_io/vendor/proc-macro2-1.0.80/src/lib.rs#1282)).
+
+### Audits should ignore criteria required by `config.toml` policy
+
+Audits should record the actual risk of the crate, even if a looser
+certification would be sufficient to pass a `config.toml` policy.
+
+For example, Chromium requires `ub-risk-2` for crates used in the browser
+process - e.g. see the following auto-generated `config.toml`
```
Comment (djkoloski) [inline auditing_standards.md:264]: Nit: remove "e.g." since we've already stated this is an example.
```suggestion
process. The following `config.toml` is auto-generated for the quote crate:
```

File: auditing_standards.md [hunk 5]
```diff
@@ -237,3 +237,76 @@ Crates with this criteria contain very dangerous unsafe rust code.
...
+
+```
+# supply-chain/config.toml:
+[policy."quote:1.0.36"]
+criteria = ["does-not-implement-crypto", "safe-to-deploy", "ub-risk-2"]
+```
+
+This means that certifying this crate as `ub-risk-2` is sufficient for `cargo
+vet` checks/presubmits to pass and `cargo vet check` will actually only ask
+to certify at the `ub-risk-2` level - see the example below:
```
Comment (djkoloski) [inline auditing_standards.md:275]: Nit: remove example clause
```suggestion
to certify at the `ub-risk-2` level:
```

File: auditing_standards.md [hunk 6]
```diff
@@ -237,3 +237,76 @@ Crates with this criteria contain very dangerous unsafe rust code.
...
+fixes all the issues that were identified in the previous, baseline audit.
+
+For example, even though
+[the 1.0.79 to 1.0.80 delta of the `proc-macro2` crate](https://chromium-review.googlesource.com/c/chromium/src/+/5453972/2..6)
+doesn't add or change any `unsafe` code, it shouldn't be recorded as
+`ub-risk-0`, because there is still `unsafe` outside of the delta
+(e.g. `unsafe { imp::Literal::from_str_unchecked(repr) }` in
+[`lib.rs` on line 1282](https://chromium-review.googlesource.com/c/chromium/src/+/5453972/2..6/third_party/rust/chromium_crates_io/vendor/proc-macro2-1.0.80/src/lib.rs#1282)).
+
+### Audits should ignore criteria required by `config.toml` policy
```
Comment (djkoloski) [inline auditing_standards.md:258]: Nit: clarity
```suggestion
### Audits should be as precise as possible

*e.g. Audits should not account for `config.toml` policy*
```

File: auditing_standards.md [hunk 7]
```diff
@@ -237,3 +237,76 @@ Crates with this criteria contain very dangerous unsafe rust code.
...
+
+Audits should record the actual risk of the crate, even if a looser
+certification would be sufficient to pass a `config.toml` policy.
```
Comment (djkoloski) [inline auditing_standards.md:261]:
```suggestion
Audits should record the most precise criteria for each crate, without
accounting for a project's `config.toml` policy. In some cases, a project's
policy may only require a crate to meet a minimum level of certification.
Although it may be less work to audit a crate at that minimum level, we
should always audit crates with the most precise criteria available. This
ensures that our audits are as shareable as possible between project
participants.
```

File: auditing_standards.md [hunk 8]
```diff
@@ -237,3 +237,76 @@ Crates with this criteria contain very dangerous unsafe rust code.
...
+    Command                             Publisher  Used By                                    Audit Size
+    cargo vet diff quote 1.0.35 1.0.36  dtolnay    syn, chromium, prost-derive, and 6 others  6 files changed, 568 insertions(+), 548 deletions(-)
+
+estimated audit backlog: 1116 lines
+
+Use |cargo vet certify| to record the audits.
+```
+
+Nevertheless, the audit should ignore the policy requirements and certify
+`quote` 1.0.36 as `ub-risk-0` because it doesn't have any `unsafe` code.
```
Comment (djkoloski) [inline auditing_standards.md:294]:
```suggestion
Although it is recommended to audit at `ub-risk-2`, the audit should always
certify `quote` 1.0.36 at the most precise ub-risk level possible. In this case,
the crate should be audited as `ub-risk-0` because it doesn't contain any
unsafe code.
```

File: auditing_standards.md [hunk 9]
```diff
@@ -237,3 +237,76 @@ Crates with this criteria contain very dangerous unsafe rust code.
...
+  or `#[cfg(target_os = "solaris")]`)
+
+**TODO**: Spell out the guidelines for auditing such conditional code.
+Consider that:
+
+* It is impossible to predict what code may be enabled by consumers of `audits.toml`.
+  For example, Chromium may rely on `audits.toml` imported from Fuchsia, even
+  though Chromium targets additional operating systems.
+* `cargo vet` has no way to indicate partial audits - see:
+  https://github.com/mozilla/cargo-vet/issues/380
```
Comment (djkoloski) [inline auditing_standards.md:312]: Can we move this to a separate PR since we don't have a consensus yet?

**Review by djkoloski — CHANGES_REQUESTED** [review]

File: auditing_standards.md (hunk 2 above)
Comment (Manishearth) [reply auditing_standards.md:243]: I believe upstream documents delta audits as being basically the New audit

Comment (anforowicz) [reply auditing_standards.md:243]: > This change assumes that doing this:
> 
> 1. Audit crate version 1 with ub-risk-3.
> 2. Audit crate delta 1->2 with ub-risk-0.
> 
> Results in crate version 2 having ub-risk-0. To my mind, this series of events should still result in version 2 having ub-risk-3

Yes, your understanding is correct.
* Version 1 is `ub-risk-3` (because of direct audit) and `ub-risk-4` (because `ub-risk-3` `implies` `ub-risk-4`)
* Edge between version 1 and version 2 delta is `ub-risk-0` (because of direct audit) and `ub-risk-1`, `ub-risk-2`, `ub-risk-3`, `ub-risk-4` (because `ub-risk-N` `implies` `ub-risk-<N+1>`)
* Version 2 is *not* `ub-risk-0` because version 1 was not `ub-risk-0` and therefore the edge doesn't apply

But... I think that:
* Having explicit guidance what to do would be desirable
* I think that stating the final state is more useful, because:
    - Saying `ub-risk-0` is misleading when the crate has `unsafe`.  In the future, when I review an audit of a delta I want to know if there is `unsafe` (so that I can evaluate if changes of safe code may impact crate-internal invariants of unchanged `unsafe` code elsewhere in the crate).  And tracing back through delta audit edges seems a bit painful.  I guess I could also patch-in each audit and grep for `unsafe`, but this is also a bit of extra work.
    - In general, a delta review of `ub-risk-2` has to consider the overall crate (and not just the delta) and therefore certifying the final state seems more correct.

File: auditing_standards.md (hunk 9 above)
Comment (anforowicz) [reply auditing_standards.md]: Done

File: auditing_standards.md (hunk 2 above)
Comment (Manishearth) [reply auditing_standards.md:243]: https://mozilla.github.io/cargo-vet/audit-entries.html#delta

Comment (Manishearth) [reply auditing_standards.md:243]: @djkoloski we've been pretty good about this in Google3: we will often just review a diff and apply a "diff review only" tag that doesn't get upstreamed as an audit. We mark things as full audits when we have looked at past audits, which is pretty easy given that we make sure we track comments. Or just look at the rest of the unsafe.

I don't really see a question of us changing this norm, given that this is in cargo vet's docs:

> The standard here is that the properties are actually preserved, not merely that that the diff doesn't obviously violate them. It is the responsibility of the auditor to acquire sufficient context to certify the former.

We'd be violating that contract if we changed this.

**Comment by anforowicz** [comment]
@Manishearth and/or @djkoloski - what are the next steps needed to finish reviewing this PR?

File: auditing_standards.md (hunk 2 above)
Comment (djkoloski) [reply auditing_standards.md:243]: Alright, I'm fine with this line of reasoning.

File: auditing_standards.md [hunk 10]
```diff
@@ -237,3 +237,67 @@ Crates with this criteria contain very dangerous unsafe rust code.
...
+
+## Criteria-agnostic guidelines
+
+### Delta audits should describe the final version
+
+In general, delta audits should always take the previous, baseline into
+account.  This means that inspecting just the source code delta should never result
+in lowering the `UB-risk-N` value from the previous audit, or downgrading `crypto-safe` to
+`does-not-contain-crypto`.  OTOH delta audits can lower the UB risk if the delta
+fixes all the issues that were identified in the previous, baseline audit.
```
Comment (djkoloski) [inline auditing_standards.md:249]: Can you add a note that since deltas only determine which criteria from the source version still apply, you need a non-delta revision to lower the UB risk.

**Review by djkoloski — CHANGES_REQUESTED** [review]
I'm fine with requiring inspection of the base audit. Just one small change and then we'll approve and merge.

File: auditing_standards.md (hunk 10 above)
Comment (anforowicz) [reply auditing_standards.md]: Done.

**Review by djkoloski — APPROVED** [review]

File: auditing_standards.md [hunk 11]
```diff
@@ -237,3 +237,86 @@ Crates with this criteria contain very dangerous unsafe rust code.
...
+`ub-risk-0`, because there is still `unsafe` outside of the delta
+(e.g. `unsafe { imp::Literal::from_str_unchecked(repr) }` in
+[`lib.rs` on line 1282](https://chromium-review.googlesource.com/c/chromium/src/+/5453972/2..6/third_party/rust/chromium_crates_io/vendor/proc-macro2-1.0.80/src/lib.rs#1282)).
+
+Another example is the improvements in `flate2` between version 1.0.27
+(`ub-risk-4` because of https://github.com/rust-lang/flate2-rs/issues/220) and
+1.0.28 (`ub-risk-3` because of https://github.com/rust-lang/flate2-rs/pull/373).
+Recording this improvement as a delta audit would be insufficient to pass `cargo
+vet check` presumits (as tested by
+[this ad-hoc Chromium CL](https://crrev.com/c/5532326)).
```
Comment (anforowicz) [inline auditing_standards.md:275]: @Manishearth, can you PTAL (since this contradicts what you've said earlier)?

In particular, I think the current situation is:

* `vet` sees that `flate2@1.0.27` is `ub-risk-4`
* `vet` sees a delta audit between `flate2@1.0.27` and `flate2@1.0.28` that declares `ub-risk-3` which also implies `ub-risk-4`
* Since `flate2@1.0.27` is *not* `ub-risk-3`, the delta edge doesn't apply / doesn't result in 1.0.28 being treated as `ub-risk-3`

OTOH, I don't have high confidence in the test I've done using the ad-hoc Chromium CL linked from the text above.  Based on my reasoning `ub-risk-4` should be okay, but it still doesn't pass/import... :-/

File: auditing_standards.md (hunk 10 above)
Comment (anforowicz) [reply auditing_standards.md]: @djkoloski, I probably should have shouted to hold off with merging because I made some substantial rewrites.  I think the new text is okay, but the changes to `manual-sources/google3-audits.toml` probably should have required a re-review... :-/  Ooops...

File: auditing_standards.md (hunk 11 above)
Comment (anforowicz) [reply auditing_standards.md:275]: > OTOH, I don't have high confidence in the test I've done using the ad-hoc Chromium CL linked from the text above. Based on my reasoning `ub-risk-4` should be okay, but it still doesn't pass/import... :-/

Oh, this is because when testing I've imported 1.0.30, but the audit only covers 1.0.28.  I've redone my tests (copy&pasting audits from google/rust-crate-audits repo into a local `audits.toml` in Chromium) and I am getting expected results (checks don't pass as `ub-risk-3` before this PR / with delta audit,  and checks pass after this PR / with non-delta audit).

So, this probably deserves a quick review, but I am quite confident that the final/merged state of this PR is correct.
